docs/digits_vae.md

```diff
@@ -0,0 +1,513 @@
+     16   return loss
+```
+and further down from this, the details of the gradient update step where the NaN is arising:
+```
+/usr/local/lib/python3.10/dist-packages/optax/tree_utils/_tree_math.py in <lambda>()
+    280       lambda g, t: (
+--> 281           (1 - decay) * (g**order) + decay * t if g is not None else None
+    282       ),
+```
+This tells us that the gradient is returning values that lead to `NaN` during the model update: typically this would come about when the gradient itself is for some reason diverging.
+
+A diverging gradient means that something with our loss function may be amiss. We previously saw `loss=NaN` at iteration 500; let's print the progress up to this point:
+
+```{code-cell}
+:id: KJ1gAh8uurVX
+:outputId: 21564ba8-2ea4-42fb-bf9a-0291a556004c
+
+model = VAE(
+  image_shape=(8, 8),
+  hidden_size=32,
+  latent_size=8,
+  rngs=nnx.Rngs(0, noise=1),
+)
+
+optimizer = nnx.Optimizer(model, optax.adam(1e-3))
+
+for epoch in range(301):
+  loss = train_step(model, optimizer, images_train)
+  if epoch % 50 == 0:
+    print(f'Epoch {epoch} loss: {loss}')
+```
+
++++ {"id": "Jk_wdvqpurTG"}
+
+It looks like our loss value is decreasing toward negative infinity until the point where the values are no longer well-represented by floating point math.
+
+At this point, we may wish to expect the values within the loss function itself to see where the diverging loss might be coming from.
+In typical Python programs you can do this by inserting either a `print` statement or a `breakpoint` in the loss function; it might look something like this:
+
+```{code-cell}
+:id: 9Klkz7qHwWia
+:outputId: e727cf56-0cc1-4256-9567-5fe96396ce45
+
+def vae_loss(model: VAE, x: jax.Array):
+  logits, mean, std = model(x)
+  kl_loss = jnp.mean(0.5 * jnp.mean(
+      -jnp.log(std ** 2) - 1.0 + std ** 2 + mean ** 2, axis=-1))
+  reconstruction_loss = jnp.mean(
+    optax.sigmoid_binary_cross_entropy(logits, x)
+  )
+  print("kl loss", kl_loss)
+  print("reconstruction loss", reconstruction_loss)
+  return reconstruction_loss + 0.1 * kl_loss
+
+model = VAE(
+  image_shape=(8, 8),
+  hidden_size=32,
+  latent_size=8,
+  rngs=nnx.Rngs(0, noise=1),
+)
+
+optimizer = nnx.Optimizer(model, optax.adam(1e-3))
+train_step(model, optimizer, images_train)
+```
+
++++ {"id": "mawnZtF8wxu9"}
+
+But here rather than printing the value, we're getting some kind of `Traced` object. You'll encounter this frequently when inspecting the progress of JAX programs: tracers are the mechanism that JAX uses to implement transformations like `jit` and `grad`, and you can read more about them in [JAX Key Concepts: Tracing](https://jax.readthedocs.io/en/latest/key-concepts.html#tracing).
+
+For our purposes, the workaround is to use another tool from the [JAX debugging](https://jax.readthedocs.io/en/latest/debugging/index.html#interactive-inspection-with-jax-debug) link above: namely `jax.debug.print`, which lets us print runtime values even when they're traced:
```

Comment:
I don't think JAX debugging is linked above?

Reply:
It's linked on line 221, but with a different link text. I could make them match – WDYT?

Reply:
Ah yeah making them match sounds good. I tried searching for "JAX debugging" to find where it was mentioned above because I didn't remember seeing it.

Reply:
text updated